File: ignite-extensions/modules/ignite-graph/works-ignite-common/src/main/java/de/kp/works/ignite/IgniteConnect.java
package de.kp.works.ignite;


import org.apache.ignite.binary.*;
import org.apache.ignite.configuration.IgniteConfiguration;

import java.util.Optional;

import org.apache.ignite.*;

public class IgniteConnect
{
	public static Ignite defaultIgnite = null; // first ignite
	
	private final Ignite ignite;
    private final String graphNS;
    
    
    public static IgniteConnect getInstance(IgniteConf conf) {
        return new IgniteConnect(conf);
    }
    
    public String graphNS() {
        return this.graphNS;
    }
    
    private Ignite ignite() {
        return ignite;
    }
  
    
    public Ignite getIgnite() {
        return this.ignite();
    }
    
    
    public IgniteCache<String, BinaryObject> getOrCreateCache(final String name) {
        final IgniteCache<String, BinaryObject> cache = IgniteUtil.MODULE$.getOrCreateCache(this.ignite(), name, this.graphNS());
        if (cache == null) {
            throw new RuntimeException("Connection to Ignited failed. Could not create cache.");
        }
        return cache;
    }
    
    public boolean cacheExists(final String cacheName) {
        return this.ignite().cacheNames().contains(cacheName);
    }
    
    public void deleteCache(final String cacheName) {
        final boolean exists = this.ignite().cacheNames().contains(cacheName);
        if (exists) {
            this.ignite().cache(cacheName).destroy();
        }
    }
    
    public Ignite getOrStart(IgniteConf cfg) {
    	Ignite ignite = null;
    	try {
    		if(defaultIgnite==null) {
    			if(Ignition.allGrids().size()==0) {
    				defaultIgnite = Ignition.start(cfg.fromFile());
    			}
    			else {
    				defaultIgnite = Ignition.allGrids().get(0);
    			}
    		}
    		ignite = Ignition.ignite(cfg.igniteName);
    		return ignite;
    	}
    	catch(Exception e) {
    		
    		try {
    			IgniteConfiguration config = cfg.fromConfig();
    			if(config==null) {
    				IgniteConf.logger.error("Can not find cfg for "+ cfg.igniteName +", use default ignite");
    				return defaultIgnite;
    			}
    			ignite = Ignition.getOrStart(config);
    			return ignite;
    		}
    		catch(Exception e2) {
    			IgniteConf.logger.error("Can not find "+ cfg.igniteName +", use default ignite", e2);
    			return defaultIgnite;
    		}
    	}
        
    }
    
    public IgniteConnect(IgniteConf conf) {
        this.graphNS = conf.igniteName;       
        this.ignite = this.getOrStart(conf);               
    }
}

File: ignite-extensions/modules/ignite-graph/works-ignite-common/src/main/scala/de/kp/works/ignite/IgniteConnect.scala
package de.kp.works.ignite

/*
 * Copyright (c) 2019 - 2021 Dr. Krusche & Partner PartG. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 * @author Stefan Krusche, Dr. Krusche & Partner PartG
 *
 */

import org.apache.ignite._
import org.apache.ignite.binary.BinaryObject

object IgniteConnect {

  private var instance:Option[IgniteConnect] = None
  private var graphNS:Option[String] = None

  def getInstance(namespace: String): IgniteConnect = {

    graphNS = Some(namespace)

    if (instance.isEmpty)
      instance = Some(new IgniteConnect(namespace))

    instance.get

  }

  def namespace:String = {
    if (graphNS.isDefined) graphNS.get else null
  }
}

class IgniteConnect(val graphNS: String) {

  private var ignite:Ignite = getOrStart()

  def getIgnite: Ignite = ignite

  /**
   * A helper method to definitely ensure that
   * an Apache Ignite node is started
   */
  def getOrCreate:Ignite = {

    if (ignite == null) ignite = getOrStart()
    ignite

  }

  def getOrCreateCache(name: String): IgniteCache[String, BinaryObject] = {
    /*
     * .withKeepBinary() must not be used here
     */
    val cache = IgniteUtil.getOrCreateCache(ignite, name, graphNS)
    if (cache == null)
      throw new Exception("Connection to Ignited failed. Could not create cache.")
    /*
     * Rebalancing is called here in case of partitioned
     * Apache Ignite caches; the default configuration,
     * however, is to use replicated caches
     * cache.rebalance().get()
     */
    
    cache
  }

  def cacheExists(cacheName: String): Boolean = {
    ignite.cacheNames.contains(cacheName)
  }

  def deleteCache(cacheName: String): Unit = {
    val exists: Boolean = ignite.cacheNames.contains(cacheName)
    if (exists) {
      ignite.cache(cacheName).destroy()
    }
  }

  /**
   * Executing Apache Spark based dataframe write
   * operations requires to have an Apache Ignite
   * node started
   */
  def getOrStart(): Ignite = {
    Ignition.getOrStart(IgniteConf.fromConfig)
  }

}
